### ISSUE TITLE
re-work Window::post_event_processing

### ISSUE THREAD

**Issue by cmyr** [linebender/druid#1603]
`post_event_processing` does a number of tasks:
- it registers new timers (all contexts)
- it sends `WidgetAdded` if a widget was added (`event`, `lifecycle` and `update`, but not in `layout`)
- it forwards animation requests (all but `layout`)
- it schedules an idle call if there are any unhandled commands, but only if a flag is set
- it updates any invalid regions accumulated during the traversal

It is also called after a bunch of different traversal methods: `layout`, `update`, `lifecycle`, and `event`.

This means there's a bunch of duplicate calling, since (for instance) an event can cause multiple lifecycle calls, there's a lifecycle call in layout, etc.

Let's think this through.

The druid event-loop can be thought of as having two 'phases'; the *event* phase and the *render* phase.

The *event* phase begins when some event, like a key-press, arrives from the system. We handle that event; we handle any commands dispatched during that event, and then we call `update`. The `update` call represents the end of the event phase, and it is guaranteed to be called every time an event arrives. At the end of the event phase, if any part of the view is marked as invalid or as needing layout, we ask the platform to call us back for the render phase.

The *render* phase begins when we get that callback. First we layout any widgets that need it, and then we paint any widgets that need it.

Given this, when should we be doing the various tasks currently performed in `post_event_processing`?

- timers:  timer state is stored in the transitive `WidgetState` object, so we need to merge these after each traversal.
- `WidgetAdded`: I believe this should always be added *immediately* after whatever event caused the widget to be added; so we should check this aggressively, likely at the end of all `event`/`lifecycle`/`update`.
- animation requests: these can be handled once, at the end of `update`.
- unhandled commands: there are two scenarios where we need to schedule an idle handler when there are pending commands: The first is when they are submitted during `update` *and there will be no render phase*, and the second is when they are submitted during the render phase.
- invalid regions: these *need* to be handled after each traversal, because they do not persist.

**conclusion**: these items should be split into three: things that need to be done after *every* traversal; things that can be done after `update` only, and things that can be done after `paint` only:

- after each traversal: timers, `WidgetAdded`, invalidation.
- after update: animation frame, schedule idle *if nothing has been invalidated*.
- after paint: idle, if there are commands.

annoyingly, timers should be added after layout, but `WidgtAdded` and invalidation should be disallowed here; however those should not be settable via the `LayoutCtx` so I  think we can just treat them the same.

**Comment by jneem** [comment]
Do you envision making druid-shell's invalidation vs animation handling clearer as part of this? (i.e. make it so that invalidation does not imply scheduling a new render pass.)

**Comment by cmyr** [comment]
Although that's related, I think it's separate.

I've spent a bunch of time going through the stuff described here, and my current thinking is that it *feels bad* but isn't quite exactly *wrong*. This also isn't describing the issue with `invalidate_and_finalize` being slightly overloaded.

**Comment by xarvic** [comment]
Since nearly all lifecycle events are called from `post_event_processing` we could think about removing `post_event_processing`  from lifecycle and enforce a strict order of Lifecycle events instead. 
(called during event) HotChanged  -> (called after event) WidgetAdded -> DisabledChanged -> BuildFocusChain -> FocusChanged  -> (called during layout) Size.
Calling `set_disabled()` from `WidgetAdded` makes sense but calling `children_changed()` from `DisabledChanged()` additionally will result in an infinite loop! All changes to events before the current event should get deleted.

**Comment by cmyr** [comment]
Yes, I agree that the recent changes around focus/enabled/disabled have made `post_event_processing` even scarier than before. I'm definitely open to trying to break this up into something into smaller methods that are called more judiciously based on the exact event that has occurred.
